### PR TITLE
Add a simple logging facility for procedures

### PR DESCRIPTION
**Pull request by boggle**

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.neo4j.kernel.impl.proc;
+
+/**
+ * LoggingService provides a logging facility for procedures. It should only be used
+ * within a procedure call
+ */
+public interface LoggingService
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java]: Is `org.neo4j.kernel.impl.proc` public API? The package name doesn't sound like it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+{
+    boolean isDebugEnabled();
+
+    void error( String message );
+    void error( String message, Object... args );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java]: These guys need javadocs. For instance, there's no way to tell how the `message` should be formatted to interpolate and format the `args`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java (hunk 1 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java]: Yep

Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java]: We're going to move these things in bulk sometime next week

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java (hunk 2 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/LoggingService.java]: We also plan to do this as a separate work item

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java [hunk 3]
```diff
@@ -828,6 +823,9 @@ private Procedures setupProcedures() throws KernelException, IOException
         procedures.registerComponent( GraphDatabaseService.class,
                 (ctx) -> dependencyResolver.resolveDependency( GraphDatabaseService.class ) );
 
+        InternalProcedureLoggingService loggingService = new InternalProcedureLoggingService( log.getUserLog( Procedures.class  ) );
+        procedures.registerComponent( LoggingService.class, (ctx) -> loggingService );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java]: Why is `LoggingService` needed? Can't we instead register `Log` or `Logger` with `log.getUserLog(Procedures.class)` as implementation?

File: integrationtests/src/test/java/org/neo4j/procedure/ProcedureIT.java [hunk 4]
```diff
@@ -229,6 +235,39 @@ public void shouldLoadBeAbleToCallProcedureWithAccessToDB() throws Throwable
...
+    @Test
+    public void shouldLogLikeThereIsNoTomorrow() throws Throwable
+    {
+        // Given
+        new JarBuilder().createJarFor( plugins.newFile( "myProcedures.jar" ), ClassWithProcedures.class );
+
+        // When
+        AssertableLogProvider logProvider = new AssertableLogProvider();
+
+        GraphDatabaseService db = new TestGraphDatabaseFactory()
```
Comment (lutovich) [inline integrationtests/src/test/java/org/neo4j/procedure/ProcedureIT.java:252]: `#shutdown()` is never called on this db

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java (hunk 3 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java]: Aren't Log and Logger internal? This was meant as just a public facade interface for use by procedure authors.
Is org.neo4j.logging part of the public api, if so, we could just hand down that, perhaps just masking out bulk logging as abusing that  may have undesirable properties.

Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java]: I assume it is public but not entirely sure. `LogProvider` is exposed from `GraphDatabaseFacadeFactory.Dependencies` interface which can be used to create `CommunityNeoServer` which is probably public.

**Comment by cleishm** [comment]
Yeah, @lutovich is right: Log and LogProvider are intended for injecting logging into Neo4j. They provide the interfaces that the internals use. No concern about the bulk parts, because it's used judiciously internally.

@nawroth is actually fixing the lack of public doc for those currently.

So suggest not adding a second interface for log injection: just use LogProvider, and definitely talk to @lutovich, @nawroth and @toblin before doing so.

**Comment by cleishm** [comment]
Oh, scratch that. Now that I read the code I realised its to provide a logging facility _to_ user code (procs). So perhaps bulk is an issue there? I'll leave it for you guys to work out and shut up now :)

**Pull request merged**